### DISCUSSION TITLE
Any Thoughts on a Framebuffer?

### DISCUSSION THREAD

**Discussion by aphistic** [oxidecomputer/hubris#484] [Ideas]
The project I'm planning to use Hubris for is going to include a display (either OLED monochrome or TFT color, I haven't decided yet) and I was wondering if there were any plans to include something like a "framebuffer" task that would interact with display drivers and provide a consistent API for rendering? Forgive me if I'm using terminology wrong, I haven't had a chance to dig into Hubris aside some of the docs yet.

I don't know what your use of Hubris is right now, but I imagined you might have some status displays you would want to include on your systems at some point. I thought maybe you'd considered that and had some thoughts on the matter? If not, I'll probably just blaze my own trails and break things as I go! :D

**Comment by labbott** [comment]
A framebuffer isn't on the official plan.  I'd be very interested to see what you come up with!

**Reply by aphistic** [reply]
Are you planning on having displays? If so, what are your thoughts on them? Would you write a per-app task/driver to handle it instead?

**Reply by cbiffle** [reply]
Our product doesn't have a display on it, so, our mainline work isn't going to involve a display. That being said, I'm a graphics nerd and I'd be more than happy to bounce ideas around here.

(Sorry for the delay in responding, afaict discussions doesn't actually notify me when things happen?)

Hubris-the-kernel is unlikely to ever have a notion of a framebuffer, simply because Hubris-the-kernel is almost entirely hardware independent and wants to stay that way (I asked it). Like other drivers, this would probably be best left in a task. A task can do more or less whatever it wants in its memory -- it might service graphics calls over IPC, or the whole UI might live in that task along with the framebuffer and just send high-level control messages out to other entities.

As far as I'm aware, the most complete project of this sort is [Artemis Everfree's port of Hubris to the PineTime watch](https://artemis.sh/2022/03/28/oxide-hubris-on-pinetime.html), which includes a SPI-interfaced display hung off the nRF52 micro.

**Reply by aphistic** [reply]
> (Sorry for the delay in responding, afaict discussions doesn't actually notify me when things happen?)

No problem! I asked @steveklabnik where the best place to talk about ideas with you all would be and he mentioned this is probably the best, but that it doesn't notify people and it might take some time for people to respond. So I wasn't expecting a quick response! :)

> Our product doesn't have a display on it, so, our mainline work isn't going to involve a display. That being said, I'm a graphics nerd and I'd be more than happy to bounce ideas around here.
> 
> Hubris-the-kernel is unlikely to ever have a notion of a framebuffer, simply because Hubris-the-kernel is almost entirely hardware independent and wants to stay that way (I asked it). Like other drivers, this would probably be best left in a task. A task can do more or less whatever it wants in its memory -- it might service graphics calls over IPC, or the whole UI might live in that task along with the framebuffer and just send high-level control messages out to other entities.

Have you had conversations internally about how you'd like to handle cases like this as far as upstreaming features? Do you think something like a framebuffer task would potentially be upstreamed even if you wouldn't use it on your own devices, or are you looking to keep your own repository rather clean of non-Oxide functionality?

This question, I guess, also coincides with different chip support as well. I'm currently working on porting Hubris and Humility to work with the RP2040 and there are things here and there (just tooling so far) that I've made small tweaks to, so I'm wondering whether I should expect to keep that in a fork or if you'd like to see it in a PR? Speaking of PRs, is there anything you don't want to see in a PR? I'm excited about contributing to the ecosystem, but I don't know what your plans are so I'm a little reluctant to make PRs to add things I'd like to see that just end up being annoying.

Getting back to framebuffers, I don't have a solid idea of what something like that would look like right now. I've never written one before (or written an OS before), so I was going to look at what others do. Right now I was thinking of something like a task that can take graphics calls over IPC (as you mention), and then maybe a common interface behind that for controller driver tasks to implement to try and abstract the actual display from the "user" code a bit. It seems like the ST7789 and SSD1306 are common controllers, so I'd probably look at what those are capable of and try to design a common API around those. 

> As far as I'm aware, the most complete project of this sort is [Artemis Everfree's port of Hubris to the PineTime watch](https://artemis.sh/2022/03/28/oxide-hubris-on-pinetime.html), which includes a SPI-interfaced display hung off the nRF52 micro.

I wasn't aware of this, but it looks like a great resource to read! Thanks for the heads up!

**Reply by mchodzikiewicz** [reply]
This feels like the hypothetical  display task could greatly benefit from already established abstraction from [embedded-graphics](https://docs.rs/embedded-graphics/latest/embedded_graphics/) - it could have a `DrawTarget` implemented for a specific display driver either by directly accessing the framebuffer memory (for example for LTDC) or via SPI / I2C / whatever task IPC for external display controllers, as an IPC the drawing task could receive a list of `Drawable`s

A question after a brief study of Hubris I don't know the answer yet is if you can pass `dyn` traits using leases, lack of it might be a show stopper here.

a nice to have would be to be able to express generic types (I guess only the pixel color type in a most recent version of the crate) for mentioned traits in the `app.toml` to make sender tasks more portable

**Reply by pokeylope** [reply]
I've been goofing around with [getting Hubris running on my Flipper Zero](https://github.com/pokeylope/hubris-flipper), which uses an ST7565 controller over SPI. I have it to the point where I can draw to the screen (using embedded-graphics), but I haven't yet figured out the interface to let other tasks interact with it.

**Reply by pokeylope** [reply]
I now have [graphics working over IPC](https://github.com/pokeylope/hubris-flipper/commit/4a91c259dc21b6aa8ca37e3fcdecbbbbc1e10099). The `display-api` driver provides a `DrawTarget` implementation that sends the commands to the display task, which then forwards the calls to its own `DrawTarget` for the actual hardware, so it should work for any device that has embedded-graphics support (although I only have the one to test it with). It's still got some rough edges (for example, error handling should be improved, and it doesn't attempt any sort of mutual exclusion), but it works fine.

**Reply by mchodzikiewicz** [reply]
This is cool!
I am a bit afraid that IPC round trip between each primitive is quite hefty overhead for any sort of higher level GUI, but that's a good start!
I could try to port your example to STM32 LTDC and see how it goes

**Reply by mchodzikiewicz** [reply]
I mentioned this discussion in an `embedded-graphics` issue about adding `flush()` method/trait for committing a drawn frame to the display - it would be useful to implement sending set of drawables through IPC

**Reply by ghost** [reply]
@cbiffle @aphistic

Consider checking out [Slint](https://github.com/slint-ui/slint).

They have demos for 2 microcontrollers; [STM32H7](https://www.youtube.com/watch?v=NNNOJJsOAis), [RP2040](https://www.youtube.com/watch?v=dkBwNocItGs)

Enjoy.

CC: @tronical
